src/servers/http_server_v2.cc

```diff
@@ -0,0 +1,1566 @@
+TRTSERVER_Error*
+HTTPServerV2Impl::Start()
+{
+  if (!worker_.joinable()) {
+    evbase_ = event_base_new();
+    htp_ = evhtp_new(evbase_, NULL);
+    evhtp_enable_flag(htp_, EVHTP_FLAG_ENABLE_NODELAY);
+    evhtp_set_gencb(htp_, HTTPServerV2Impl::Dispatch, this);
+    evhtp_use_threads_wexit(htp_, NULL, NULL, thread_cnt_, NULL);
+    evhtp_bind_socket(htp_, "0.0.0.0", port_, 1024);
+    // Set listening event for breaking event loop
+    evutil_socketpair(AF_UNIX, SOCK_STREAM, 0, fds_);
+    break_ev_ = event_new(evbase_, fds_[0], EV_READ, StopCallback, evbase_);
+    event_add(break_ev_, NULL);
+    worker_ = std::thread(event_base_loop, evbase_, 0);
+    return nullptr;
+  }
+
+  return TRTSERVER_ErrorNew(
+      TRTSERVER_ERROR_ALREADY_EXISTS, "HTTP V2 server is already running.");
+}
+
+TRTSERVER_Error*
+HTTPServerV2Impl::Stop()
+{
+  if (worker_.joinable()) {
+    // Notify event loop to break via fd write
+    send(fds_[1], &evbase_, sizeof(event_base*), 0);
+    worker_.join();
+    event_free(break_ev_);
+    evutil_closesocket(fds_[0]);
+    evutil_closesocket(fds_[1]);
+    evhtp_unbind_socket(htp_);
+    evhtp_free(htp_);
+    event_base_free(evbase_);
+    return nullptr;
+  }
+
+  return TRTSERVER_ErrorNew(
+      TRTSERVER_ERROR_UNAVAILABLE, "HTTP V2 server is not running.");
+}
+
+void
+HTTPServerV2Impl::StopCallback(int sock, short events, void* arg)
+{
+  struct event_base* base = (struct event_base*)arg;
+  event_base_loopbreak(base);
+}
+
+void
+HTTPServerV2Impl::Dispatch(evhtp_request_t* req, void* arg)
+{
+  (static_cast<HTTPServerV2Impl*>(arg))->Handle(req);
+}
+
+
+// Handle HTTP requests to inference server APIs
+class HTTPAPIServerV2 : public HTTPServerV2Impl {
+ public:
+  explicit HTTPAPIServerV2(
+      const std::shared_ptr<TRTSERVER_Server>& server,
+      const std::shared_ptr<nvidia::inferenceserver::TraceManager>&
+          trace_manager,
+      const std::shared_ptr<SharedMemoryManager>& shm_manager,
+      const std::vector<std::string>& endpoints, const int32_t port,
+      const int thread_cnt)
+      : HTTPServerV2Impl(port, thread_cnt), server_(server),
+        trace_manager_(trace_manager), shm_manager_(shm_manager),
+        allocator_(nullptr), api_regex_(R"(/v2/(health|models)(.*))"),
+        health_regex_(R"(/(live|ready))"),
+        model_regex_(R"(/([^/]+)(?:/version/([0-9]+))?(/infer|/ready)?)")
+  {
+    TRTSERVER_Error* err = TRTSERVER_ServerId(server_.get(), &server_id_);
+    if (err != nullptr) {
+      server_id_ = "unknown:0";
+      TRTSERVER_ErrorDelete(err);
+    }
+
+    FAIL_IF_ERR(
+        TRTSERVER_ResponseAllocatorNew(
+            &allocator_, InferResponseAlloc, ResponseRelease),
+        "creating response allocator");
+  }
+
+  ~HTTPAPIServerV2()
+  {
+    LOG_TRTSERVER_ERROR(
+        TRTSERVER_ResponseAllocatorDelete(allocator_),
+        "deleting response allocator");
+  }
+
+  //
+  // AllocPayload
+  //
+  // Simple structure that carries the userp payload needed for
```

Comment:
AllocPayload and InferResponseAlloc should be move to a common set of classes/functions that can be used by both HTTP and GRPC servers. They are mostly identical except for the little code needed for difference between protobuf message and evbuffer. You should break that out into a virtual function or similar so that most of this code can be shared. Please work with @tanmayv25 on this. It can be done as a separate PR to this one.

Reply:
Will do

Reply:
Did this get filed?

Reply:
Just filed it